src/cache/cache.service.ts

```diff
@@ -1,128 +1,234 @@
       );
 
-      return { headers, data } as T;
-    } catch (error) {
-      return this.cacheDefaultValue;
-    }
-  }
+      if (allContents.length < 1) {
+        return defaultValue;
+      }
 
-  private async saveCacheToFiles(): Promise<void> {
-    if (!this.cache) throw new Error('Cache is not set');
+      // checking that all headers are equal
+      const allHeadersEqual = allContents.every(
+        (content, i, arr) =>
+          JSON.stringify(arr[0]?.headers ?? '') ===
+          JSON.stringify(content.headers),
+      );
 
-    const { headers, data } = this.cache;
+      if (!allHeadersEqual) {
+        throw new CacheError('Cache headers are not equal');
+      }
 
-    const dirPath = await this.getCacheDirPath();
-    await mkdir(dirPath, { recursive: true });
+      // checking that all stats equal
+      const allStatsAreEqual = allContents.every(
+        (content, i, arr) =>
+          (arr[0]?.stats?.dataTotalLength ?? 0) ===
+          content.stats.dataTotalLength,
+      );
 
-    await this.deleteCacheFiles();
+      if (!allStatsAreEqual) {
+        throw new CacheError('Cache stats are not equal');
+      }
 
-    const totalBatches = Math.ceil(data.length / this.cacheBatchSize);
+      const headers: Headers = allContents[0].headers;
+      const stats = allContents[0].stats;
 
-    for (let batchIndex = 0; batchIndex < totalBatches; batchIndex++) {
-      const from = batchIndex * this.cacheBatchSize;
-      const to = (batchIndex + 1) * this.cacheBatchSize;
-      const batchedData = data.slice(from, to);
+      const data: Data[] = allContents.reduce<Data[]>(
+        (accumulator, content) => accumulator.concat(content.data),
+        [],
+      );
 
-      const filePath = join(dirPath, this.getCacheFileName(batchIndex));
-      await writeFile(filePath, JSON.stringify({ headers, data: batchedData }));
+      if (stats.dataTotalLength !== data.length) {
+        throw new CacheError(
+          'Headers data length does not match real data length',
+        );
+      }
+
+      return { headers, data };
+    } catch (error) {
+      this.logger?.warn('Deposit events cache will be set to a default value', {
+        error,
+      });
+
+      return defaultValue;
     }
   }
 
-  private async deleteCacheFiles(): Promise<void> {
-    try {
-      const filePaths = await this.getCacheFilePaths();
-      await Promise.all(filePaths.map(async (filePath) => unlink(filePath)));
-    } catch (error) {}
+  private async readCacheFile(
+    filePath: string,
+    valueType: z.ZodType<{ headers: Headers; data: Data[]; stats: Stats }>,
+  ): Promise<{ headers: Headers; data: Data[]; stats: Stats }> {
+    const content = await readFile(filePath, 'utf-8');
+    const parsed: Json = JSON.parse(content);
+
+    const res = valueType.safeParse(parsed);
+
+    if (res.success) {
+      return res.data;
+    }
+
+    throw new CacheError(
+      `Cache file [${basename(filePath)}] integrity error`,
+      res.error,
+    );
```

Comment:
i would suggest to follow one code style and throw the error in the condition instead of the early exit

Reply:
I would suggest putting this parsing in a separate method and not passing `valueType` through multiple methods via arguments, but getting it from `this.cacheValueType` in place

Reply:
I agree on throwing error on condition. Also I think that passing `valueType` thought multiple methods is important here, because `valueType` argument here is not the same as `this.cacheValueType` that is used in class variable. Argument `valueType` includes `stats`, and `this.cacheValueType` not.

Reply:
UPD: re-checked again, the second sentence is not true, `this.cacheValueType` and `valueType` argument are the same